pkg/template/registry/templateinstance/strategy.go

```diff
@@ -96,8 +98,34 @@ func (s *templateInstanceStrategy) ValidateUpdate(ctx apirequest.Context, obj, o
 		return field.ErrorList{field.InternalError(field.NewPath(""), errors.New("user not found in context"))}
 	}
 
-	templateInstance := obj.(*templateapi.TemplateInstance)
-	oldTemplateInstance := old.(*templateapi.TemplateInstance)
+	// Decode Spec.Template.Objects on both obj and old to Unstructureds.  This
```

Comment:
@deads2k or @liggitt ptal?

The Spec.Objects field of a v1.Template is a []RawExtension.  The garbage collector gets TemplateInstances (which wrap Templates) into Unstructureds, then posts them back to update the finalisers.  In this case the new Objects field received by the API server is semantically identical to the old one, but often the serialisation is different.  We were previously rejecting the update because of this.  Is this solution OK or is there a better alternative you can suggest?

Reply:
can you give an example of this? I'd expect perfect fidelity when using unstructured objects.

Reply:
If you deserialise to Unstructured and re-serialise, it doesn't guarantee to maintain the order of keys in a json object.  This was the example I found:

"old" value (fetched from etcd)
`{"kind":"Secret","apiVersion":"v1","metadata":{"name":"secret1","creationTimestamp":null}}`

"new" value (submitted by GC, been through the Unstructured mangler)
`{"apiVersion":"v1","kind":"Secret","metadata":{"creationTimestamp":null,"name":"secret1"}}`

Reply:
> The Spec.Objects field of a v1.Template is a []RawExtension. The garbage collector gets TemplateInstances (which wrap Templates) into Unstructureds, then posts them back to update the finalisers. In this case the new Objects field received by the API server is semantically identical to the old one, but often the serialisation is different. We were previously rejecting the update because of this. Is this solution OK or is there a better alternative you can suggest?

I'd open an issue and tag @ironcladlou or fix GC yourself to issue patches instead of updates.

Do you know what is wrong with the semantic equality check that causes it to fail on reordered json fields?  https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/registry/rbac/helpers.go#L47

Reply:
> Do you know what is wrong with the semantic equality check that causes it to fail on reordered json fields?

They're `[]byte` blobs where they're compared, not `map[string]interface{}`